### PR TITLE
SOF-1990 A/B Roll for TriCaster

### PR DESCRIPTION
**Pull request by LindvedKrvang**
Implement Server A/B roll for TriCaster

### CODE REVIEW

File: src/blueprints/tv2/tv2-on-timeline-generate-service.ts [hunk 1]
```diff
@@ -281,6 +293,47 @@ export class Tv2OnTimelineGenerateService implements BlueprintOnTimelineGenerate
...
+      return
+    }
+    const meEffectContent: TriCasterMixEffectEffectModeContent = triCasterMeTimelineObject.content.me
+    for (const key of Object.keys(meEffectContent.layers)) {
+      const layer: TriCasterLayer | undefined = meEffectContent.layers[key as TriCasterLayerName]
+      if (!layer || !layer.input || layer.input !== this.prefixSourceInputWithTriCasterPrefix(A_B_SOURCE_INPUT_PLACEHOLDER)) {
+        continue
+      }
+      layer.input = this.prefixSourceInputWithTriCasterPrefix(mediaPlayer.videoMixerSource)
+    }
```
Comment (KvelaGorrrrnio) [inline src/blueprints/tv2/tv2-on-timeline-generate-service.ts]: The key is only used to extract the layer value. A level of abstraction can be removed by using `Object.values` to work directly with the values or `Array.map` to iterate over the list. E.g.
```typescript
  const tricasterABSourceInputPlaceholder: string = this.prefixSourceInputWithTriCasterPrefix(A_B_SOURCE_INPUT_PLACEHOLDER)
  meEffectContent.layers = meEffectContent.layers
    .map(layer => {
      if (layer.input !== tricasterABSourceInputPlaceholder) {
        return layer
      }
      return { ...layer, input: this.prefixSourceInputWithTriCasterPrefix(mediaPlayer.videoMixerSource) }
    })
```
or
```typescript
  const tricasterABSourceInputPlaceholder: string = this.prefixSourceInputWithTriCasterPrefix(A_B_SOURCE_INPUT_PLACEHOLDER)
  for (const layer of Object.values(meEffectContent.layers)) {
     if (layer.input !== tricasterABSourceInputPlaceholder) {
       continue
     }
     layer.input = this.prefixSourceInputWithTriCasterPrefix(mediaPlayer.videoMixerSource)
  }

File: src/blueprints/tv2/tv2-on-timeline-generate-service.ts [hunk 2]
```diff
@@ -281,6 +293,47 @@ export class Tv2OnTimelineGenerateService implements BlueprintOnTimelineGenerate
...
+
+  private prefixSourceInputWithTriCasterPrefix(sourceInput: number): TriCasterInputName {
+    return `input${sourceInput}`
+  }
+
+  private updateTriCasterLookaheadWithMediaPlayer(timelineObject: Tv2BlueprintTimelineObject, mediaPlayer: Tv2MediaPlayer): void {
+    if (timelineObject.content.deviceType !== DeviceType.TRICASTER || timelineObject.content.type !== TriCasterType.MIX_OUTPUT) {
+      return
+    }
+    const triCasterMixOutputTimelineObject: TriCasterMixOutputTimelineObject = timelineObject as TriCasterMixOutputTimelineObject
```
Comment (KvelaGorrrrnio) [inline src/blueprints/tv2/tv2-on-timeline-generate-service.ts]: If the conditions are given a signature, the type assumptions can be removed. E.g. by creating a type guard method that uses type overloading to specify the type based on the content type:

```typescript
  private isTricasterTimelineObject(timelineObject: Tv2BlueprintTimelineObject, contentType: TriCasterType.MIX_OUTPUT): timelineObject is TriCasterMixOutputTimelineObject
  private isTricasterTimelineObject(timelineObject: Tv2BlueprintTimelineObject, contentType: TriCasterType.ME): timelineObject is TriCasterMixEffectTimelineObject
  private isTricasterTimelineObject(timelineObject: Tv2BlueprintTimelineObject, contentType: TriCasterType): timelineObject is TriCasterMixOutputTimelineObject | TriCasterMixEffectTimelineObject {
    return timelineObject.content.deviceType === DeviceType.TRICASTER  && timelineObject.content.type === contentType
  }
```

Then this method can be boiled down to:
```typescript
  private updateTriCasterLookaheadWithMediaPlayer(timelineObject: Tv2BlueprintTimelineObject, mediaPlayer: Tv2MediaPlayer): void {
    if (!this.isTricasterTimelineObject(timelineObject, TriCasterType.MIX_OUTPUT)) {
      return
    }
    timelineObject.content.source = this.prefixSourceInputWithTriCasterPrefix(mediaPlayer.videoMixerSource)
  }
```

**Review by RasmusAlbrektsen — APPROVED** [review]

File: src/blueprints/tv2/tv2-on-timeline-generate-service.ts (hunk 1 above)
Comment (LindvedKrvang) [reply src/blueprints/tv2/tv2-on-timeline-generate-service.ts]: Updated to your latter suggestion.

File: src/blueprints/tv2/tv2-on-timeline-generate-service.ts (hunk 2 above)
Comment (LindvedKrvang) [reply src/blueprints/tv2/tv2-on-timeline-generate-service.ts]: Nice suggestion :)
I have updated both methods to use this.

File: src/blueprints/tv2/tv2-on-timeline-generate-service.ts [hunk 3]
```diff
@@ -281,6 +295,51 @@ export class Tv2OnTimelineGenerateService implements BlueprintOnTimelineGenerate
     }
   }
 
+  private updateTriCasterProgramWithMediaPlayer(timelineObject: Tv2BlueprintTimelineObject, mediaPlayer: Tv2MediaPlayer): void {
+    if (timelineObject.content.deviceType !== DeviceType.TRICASTER || timelineObject.content.type !== TriCasterType.ME) {
+      return
+    }
+    const triCasterMeTimelineObject: TriCasterMixEffectTimelineObject = timelineObject as TriCasterMixEffectTimelineObject
+    if (triCasterMeTimelineObject.content.me.type !== TriCasterMixEffectContentType.PROGRAM) {
```
Comment (KvelaGorrrrnio) [inline src/blueprints/tv2/tv2-on-timeline-generate-service.ts]: Can't `isTriCasterTimelineObject` be used for both conditions here?

Comment (LindvedKrvang) [reply src/blueprints/tv2/tv2-on-timeline-generate-service.ts]: Updated

**Review by KvelaGorrrrnio — APPROVED** [review]

**Pull request merged**